### PR TITLE
Jormungandr: first draft for realtime proxy handling

### PR DESCRIPTION
**Pull request by antoine-de**
first draft for timeo realtime proxy handling

for first review, still a lots of tests need to be added, and the code need to be strengthen up

[Review on Reviewable](https://reviewable.io/reviews/canaltp/navitia/1413)

### CODE REVIEW

**Comment by xlqian** [comment]
*[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 85 \[r1\]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZdg0bJdPvNSkhH25c:-KAZdg0bJdPvNSkhH25d:1145944208) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L85)):*
return something? 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)*

**Comment by xlqian** [comment]
_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 90 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZegda1yiLq_Fg5KU1:-KAZegda1yiLq_Fg5KU2:-910487995) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L90)):_
the first test condition is not necessary 

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

**Comment by antoine-de** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 85 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZdg0bJdPvNSkhH25c:-KAZgUweNpvi2uOYT1L0:-1869267103) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L85)):_
yes the passages, why ?

---

_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 90 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZegda1yiLq_Fg5KU1:-KAZgP3P8150PV0gPWGJ:-452246526) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L90)):_
why ?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

**Comment by xlqian** [comment]
_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 126 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZgb2VaS0J8BuiB4DF:-KAZgb2VaS0J8BuiB4DG:-938218027) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L126)):_
`not all([stop, line, route])`

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

**Comment by antoine-de** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 126 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZgb2VaS0J8BuiB4DF:-KAZhFPULxrgtmWUZLbu:-510466975) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L126)):_
:+1: 

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

**Comment by xlqian** [comment]
_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 138 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZhfUsEgdom0zGAxgB:-KAZhfUtVZAY9Y64Uby4:1493218975) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L138)):_
minor suggestion:

`stop_id_url = ("StopDescription=?" 
                      "StopTimeoCode={stop}"
                      "&LineTimeoCode={line}" 
                      "&Way={route}" 
                      "&NextStopTimeNumber={count}" ).format(
                   stop=stop,
                   line=line,
                   route=route,
                   count='5',  # TODO better pagination
)`

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

**Comment by xlqian** [comment]
_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 90 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZegda1yiLq_Fg5KU1:-KAZiLNkQSBuosm-ZYLR:362899981) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L90)):_
if you get `[]` for your st_responses, then `len(st_responses) != 1` is true too... Am I missing something?

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

**Comment by xlqian** [comment]
_[source/jormungandr/jormungandr/realtime_schedule/timeo.py, line 85 [r1]](https://reviewable.io:443/reviews/canaltp/navitia/1413#-KAZdg0bJdPvNSkhH25c:-KAZilSJIQ-zUmhiAmcO:-1361132789) ([raw file](https://github.com/canaltp/navitia/blob/8fb2b324a1d0765d572e4cf4391b934fa6fe0f5a/source/jormungandr/jormungandr/realtime_schedule/timeo.py#L85)):_
Sorry, plz ignored this comment... 

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1413)_

File: source/jormungandr/jormungandr/realtime_schedule/timeo.py [hunk 1]
```diff
@@ -37,21 +64,85 @@ class Timeo(RealtimeProxy):
...
-        #TODO implem
-        return ""
+    def get_passages(self, timeo_resp):
+        logging.getLogger(__name__).debug('timeo response: {}'.format(timeo_resp))
+
+        st_responses = timeo_resp.get('StopTimesResponse')
+        # by construction there should be only one StopTimesResponse
+        if not st_responses or len(st_responses) != 1:
+            logging.getLogger(__name__).warning('invalid timeo response: {}'.format(timeo_resp))
+            return []
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/realtime_schedule/timeo.py]: sometime you return [], sometime None, what is the difference between the two?

**Comment by kinnou02** [comment]
Jenkins, test this please.

**Comment by kinnou02** [comment]
Jenkins, test this please.

**Comment by kinnou02** [comment]
Jenkins, test this please.

**Comment by kinnou02** [comment]
Jenkins, test this please.

**Comment by antoine-de** [comment]
Jenkins, test this please

**Comment by kinnou02** [comment]
Jenkins, test this please.

**Comment by kinnou02** [comment]
Jenkins, test this please.

File: source/jormungandr/jormungandr/realtime_schedule/tests/realtime_proxy_manager_test.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+def realtime_proxy_creation_test():
+    """
+    simple timeo proxy creation
+    """
+    config = [{
+                  'id': 'proxy_id',
+                  'class': 'jormungandr.realtime_schedule.timeo.Timeo',
+                  'args': {
+                      "timezone": "Europe/Paris",
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/tests/realtime_proxy_manager_test.py]: why double quote here?

File: source/jormungandr/jormungandr/realtime_schedule/tests/timeo_test.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    def get(self, url, *args, **kwargs):
+        return MockResponse(json.loads(self.responses[url]))
+
+
+def _dt(dt_to_parse="00:00", year=2016, month=2, day=7):
+    """
+    small helper to ease the reading of the tests
+    >>> _dt("8:15")
+    datetime.datetime(2016, 5, 6, 8, 15)
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/tests/timeo_test.py]: Day 6? with a default value of 7? I don't get it...

File: source/jormungandr/jormungandr/realtime_schedule/timeo.py [hunk 4]
```diff
@@ -37,21 +53,101 @@ class Timeo(RealtimeProxy):
...
+                                                          base_params=base_params,
+                                                          stop_id=stop_id_url)
+
+        return url
 
-        return self.get_passages(r.json())
+    def _get_dt(self, hour_str):
+        hour = _to_duration(hour_str)
+        # we then have to complete the hour with the date to have a datetime
+        # Note: we use now() and not utc_now() because we want a local time, the same used by timeo
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/realtime_schedule/timeo.py]: unsync comment? I can't find `now()`

File: source/jormungandr/jormungandr/realtime_schedule/tests/realtime_proxy_manager_test.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                      }
+                  }
+              }]
+
+    RealtimeProxyManager(config)  # should raise an Exception
+
+
+def multi_proxy_creation_test():
+    """
+    2 bad and one good proxy, the good one should be create
```
Comment (pbougue) [inline source/jormungandr/jormungandr/realtime_schedule/tests/realtime_proxy_manager_test.py]: minor : `created`

**Comment by antoine-de** [comment]
Jenkins, test this please

**Comment by antoine-de** [comment]
Jenkins, test this please

File: source/jormungandr/jormungandr/realtime_schedule/timeo.py [hunk 6]
```diff
@@ -28,8 +28,25 @@
 # IRC #navitia on freenode
 # https://groups.google.com/d/forum/navitia
 # www.navitia.io
+from flask import logging
+import pytz
 import requests as requests
 from jormungandr.realtime_schedule.realtime_proxy import RealtimeProxy
+from jormungandr.schedule import NextRTPassage
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/realtime_schedule/timeo.py]: realtime

**Comment by kinnou02** [comment]
:+1:

**Pull request merged**